src/native/s3select/s3select_napi.cpp

```diff
@@ -0,0 +1,279 @@
+                SetError("failed to select from json");
+            }
+        }
+
+        /*std::cout << "select res = " << this->select << std::endl;
+        std::cout.flush();*/
+    }
+
+    void OnOK() override
+    {
+        Napi::Env env = Env();
+        // in case of empty select result (ie, no rows in current buffer matched sql condition), return null
+        if (_select.empty()) {
+            _deferred.Resolve(env.Null());
+            return;
+        }
+
+        Napi::Object result = Napi::Object::New(env);
+        Napi::Value select_buf = Napi::Buffer<char>::/*New*/ Copy(env, _select.data(), _select.length());
+        result.Set("select", select_buf);
+
+        uint32_t prelude[3];
+        int32_t prelude_crc, message_crc, prelude_crc_BE;
+        prelude[0] = htonl(_select.length() + _headers_len + 16);
+        prelude[1] = htonl(_headers_len);
+        prelude_crc = crc32_gzip_refl_base(0, (uint8_t*)prelude, 8);
+        result.Set("prelude_crc", prelude_crc);
+        prelude_crc_BE = htonl(prelude_crc);
+        /*std::cout << "sz = " << sizeof(headers_bytes) << ", native prelude_crc = " << prelude_crc << ", prelude = " << std::hex << std::setfill('0') << std::setw(2) << prelude[0] << prelude[1];
+        std::cout << std::endl;
+        std::cout.flush();*/
+        message_crc = crc32_gzip_refl_base(prelude_crc, (uint8_t*)&prelude_crc_BE, 4);
+        message_crc = crc32_gzip_refl_base(message_crc, const_cast<uint8_t*>(_headers_bytes), _headers_len);
+        message_crc = crc32_gzip_refl_base(message_crc, (uint8_t*)_select.data(), _select.length());
+        result.Set("message_crc", message_crc);
+
+        _deferred.Resolve(result);
+    }
+
+    void OnError(Napi::Error const& error) override
+    {
+        Napi::Env env = Env();
+        auto obj = error.Value();
+        _deferred.Reject(obj);
+    }
+};
+
+class S3SelectNapi : public Napi::ObjectWrap<S3SelectNapi>
+{
+
+public:
+    static Napi::Object Init(Napi::Env env, Napi::Object exports);
+    S3SelectNapi(const Napi::CallbackInfo& info);
+    Napi::Value Write(const Napi::CallbackInfo& info);
+    Napi::Value Flush(const Napi::CallbackInfo& info);
+    ~S3SelectNapi();
+
+private:
+    static Napi::FunctionReference constructor;
+    Napi::ObjectReference _args_ref;
+    std::string input_format;
+    s3selectEngine::s3select s3select;
+    s3selectEngine::csv_object* csv_object = nullptr;
+    s3selectEngine::json_object* json_object = nullptr;
+    const uint8_t* headers_buf;
+    uint32_t headers_len;
+};
+
+Napi::Value
+S3SelectNapi::Write(const Napi::CallbackInfo& info)
+{
+    Napi::Buffer<char> buffer = info[0].As<Napi::Buffer<char>>();
+    SelectWorker* worker = new SelectWorker(
+        info,
+        *this,
+        csv_object,
+        json_object,
+        buffer.Data(),
+        buffer.Length(),
+        false,
+        input_format,
+        headers_buf,
+        headers_len);
```

Comment:
To make the headers buffer be safely ref/unref in the worker ctor/dtor with _args_ref you can just send it here to the function (and also in flush):
```suggestion
    Napi::Buffer<char> headers = info[0].As<Napi::Buffer<char>>();
    Napi::Buffer<char> buffer = info[1].As<Napi::Buffer<char>>();
    SelectWorker* worker = new SelectWorker(
        info,
        *this,
        csv_object,
        json_object,
        buffer.Data(),
        buffer.Length(),
        false,
        input_format,
        headers.Data(),
        headers.Length());
```

Reply:
Oh oops I just saw you added _args_ref to the S3SelectNapi object too, so that works as well. Resolving.